src/coreclr/jit/optimizer.cpp

```diff
@@ -4575,6 +4575,272 @@ PhaseStatus Compiler::optUnrollLoops()
+//
+// For example:
+// ------------ BB03 [009..00D) -> BB05 (cond), preds={BB02} succs={BB04,BB05}
+// STMT00004
+//   *  JTRUE     void   $VN.Void
+//   \--*  GE        int    $102
+//      +--*  LCL_VAR   int    V02
+//      \--*  CNS_INT   int    7 $46
+//
+// ------------ BB04 [00D..010), preds={BB03} succs={BB05}
+// STMT00005
+//   *  ASG       int    $VN.Void
+// +--*  LCL_VAR   int    V00 arg0
+// \--*  CNS_INT   int    5 $47
+//
+// The JTRUE statement in the first block is replace with a SELECT node in the
+// assignment in the second block. If the result of the SELECTs op1 is true,
+// then op2 is passed to the assignment. Otherwise op3 is used. For example:
+//
+// ------------ BB03 [009..00D), preds={BB02} succs={BB04}
+// STMT00004
+//   *  NOP       void
+//
+// ------------ BB04 [00D..010), preds={BB03} succs={BB05}
+// STMT00005
+//   *  ASG       int    $VN.Void
+//   +--*  LCL_VAR   int    V00 arg0
+//   \--*  SELECT    int
+//      +--*  LT        int    $102
+//      |  +--*  LCL_VAR   int    V02
+//      |  \--*  CNS_INT   int    7 $46
+//      +--*  CNS_INT   int    5 $47
+//      \--*  LCL_VAR   int    V00
+//
+//
+bool Compiler::optIfConvert(BasicBlock* block)
+{
+#ifndef TARGET_ARM64
+    return false;
+#else
+
+    // Don't optimise the block if it is inside a loop
+    // When inside a loop, branches are quicker than selects.
+    // Detect via the block weight as that will be high when inside a loop.
+    if (block->getBBWeight(this) > BB_UNITY_WEIGHT)
+    {
+        return false;
+    }
+
+    // Does the block end by branching via a JTRUE after a compare?
+    if (block->bbJumpKind != BBJ_COND || block->NumSucc() != 2)
+    {
+        return false;
+    }
+
+    // Verify the test block ends with a conditional that we can manipulate.
+    GenTree* last = block->lastStmt()->GetRootNode();
+    noway_assert(last->OperIs(GT_JTRUE));
+    if (!last->gtGetOp1()->OperIsCmpCompare() || (last->gtFlags & GTF_SIDE_EFFECT) != 0 ||
+        (last->gtGetOp1()->gtFlags & GTF_COLON_COND) != 0)
+    {
+        return false;
+    }
+
+    BasicBlock* middleBlock = block->bbNext;
+    GenTree*    asgNode     = nullptr;
+    Statement*  asgStmt     = nullptr;
+
+    // Check the block is followed by a block or chain of blocks that only contain NOPs and
+    // a single ASG statement. The destination of the final block must point to the same as the
+    // true path of the JTRUE block.
+    bool foundMiddle = false;
+    while (!foundMiddle)
+    {
+        noway_assert(middleBlock != nullptr);
+        BasicBlock* const middleNext = middleBlock->GetUniqueSucc();
+
+        if (middleBlock->bbNext == block->bbJumpDest)
+        {
+            // This is our final middle block.
+            foundMiddle = true;
+        }
+
+        // Check that we have linear flow and are still in the same EH region
+        //
+        if (middleBlock->NumSucc() != 1 || middleBlock->bbJumpKind != BBJ_NONE)
+        {
+            return false;
```

Comment:
The check for `NumSucc() == 1` should be redundant. Instead, you should (up above) check if `middleNext == nullptr`.

And if you generalize your flow transformation below, you can remove the check for `BBJ_NONE` here and handle the `BBJ_ALWAYS` case too.